internal/eventing/sink/http/http.go

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	sinkimpl "github.com/noctarius/timescaledb-event-streamer/internal/eventing/sink"
+	config "github.com/noctarius/timescaledb-event-streamer/spi/config"
+	"github.com/noctarius/timescaledb-event-streamer/spi/encoding"
+	"github.com/noctarius/timescaledb-event-streamer/spi/schema"
+	"github.com/noctarius/timescaledb-event-streamer/spi/sink"
+	"net/http"
+	"time"
+)
+
+func init() {
+	sinkimpl.RegisterSink(config.Http, newHttpSink)
+}
+
+func basicAuth(
+	username, password string,
+) string {
+	auth := username + ":" + password
+	return base64.StdEncoding.EncodeToString([]byte(auth))
+}
+
+type httpSink struct {
+	client  *http.Client
+	encoder *encoding.JsonEncoder
+	address *string
+	headers *http.Header
+}
+
+func newHttpSink(
+	c *config.Config,
+) (sink.Sink, error) {
+
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	if config.GetOrDefault(c, config.PropertyHttpTlsEnabled, false) {
+		transport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: config.GetOrDefault(
+				c, config.PropertyHttpTlsSkipVerify, false,
+			),
+			ClientAuth: config.GetOrDefault(
+				c, config.PropertyHttpTlsClientAuth, tls.NoClientCert,
+			),
+		}
+	}
+
+	address := config.GetOrDefault(c, config.PropertyHttpUrl, "http://localhost:80")
+	headers := make(http.Header)
+
+	authenticationType := config.GetOrDefault(c, config.PropertyHttpAuthenticationType, "none")
+	switch config.HttpAuthenticationType(authenticationType) {
+	case config.BasicAuthentication:
+		{
+			headers.Add("Authorization",
+				fmt.Sprintf("Basic %s",
+					basicAuth(config.GetOrDefault(c, config.PropertyHttpBasicAuthenticationUsername, ""),
+						config.GetOrDefault(c, config.PropertyHttpBasicAuthenticationPassword, ""),
+					),
+				),
+			)
+		}
+	case config.HeaderAuthentication:
+		{
+			headers.Add(config.GetOrDefault(c, config.PropertyHttpHeaderAuthenticationHeaderName, ""),
+				config.GetOrDefault(c, config.PropertyHttpHeaderAuthenticationHeaderValue, ""),
+			)
+		}
+	case config.NoneAuthentication:
+		{
+		}
+	default:
+		{
+			return nil, fmt.Errorf("http AuthenticationType '%s' doesn't exist", authenticationType)
+		}
+	}
+
+	return &httpSink{
+		client:  &http.Client{Transport: transport},
+		encoder: encoding.NewJsonEncoderWithConfig(c),
+		address: &address,
+		headers: &headers,
+	}, nil
+}
+
+func (h *httpSink) Start() error {
+	return nil
+}
+
+func (h *httpSink) Stop() error {
+	h.client.CloseIdleConnections()
+	return nil
+}
+
+func (h *httpSink) Emit(
+	_ sink.Context, _ time.Time, topicName string, key, envelope schema.Struct,
+) error {
+	delete(envelope, "schema")
```

Comment:
I think this mainly depends on where you'd like to take the project. Would you like to have everything or almost everything as a configurable parameter?
If so then yeah it would make sense to be able to turn on/off the sending of the schema from the config file.

I can also see a world where the package always emits everything and it's up to the receiver to discard whatever is not needed or if one would like to avoid unnecessary traffic then one could transform the payload with a plugin before emitting.

I think you probably have better things to work on in this project then having to make `schema.Struct` be optional throughout all the sinks just because this one sink and the `stdout` one does not send it.

I have to be honest and tell you that I've started the http sink by copying the `stdout` one and that one [starts the emitting process by deleting the schema](https://github.com/noctarius/timescaledb-event-streamer/blob/main/internal/eventing/sink/stdout/stdout.go#L39).

For now I think it'd be best if all the sinks behaved the same way (aside from maybe the `stdout` one, see above), not just because of consistency, but because if you decide to make the schema sending optional you'll have an easier time refactoring and not having to wade through code just to find out if a particular sink decides to omit schema by default.

I've pushed a commit that adds back in the schema for this sink as well. I'll use a plugin to transform the payload before sending to the sink anyway, so I'll just omit it there.

Reply:
Yeah I agree. I think it could be interesting since the schema creates quite some overhead and traffic. I also think it may be better to provide the option to create a separate stream of schema data, which could be routed to a schema registry, but I never used one before so I wouldn't have enough insight into it.

I think for now we can just keep it in and make it up to the receiver to discard it. For traffic reasons it may be interesting to eventually provide the option. Tbh, even stdout should support it, since the idea (next to testing) was to pipe the output somewhere else. The reason why you can send any log message to stderr 😁

Reply:
> [...]  since the schema creates quite some overhead and traffic.  [...] For traffic reasons it may be interesting to eventually provide the option

Yeah, this is probably the biggest argument for making it optional. 

> I think for now we can just keep it in and make it up to the receiver to discard it.

What are your thoughts on discarding it with a Plugin before sending it? We're streaming anywhere between 10k - 100k per second with `timescaledb-event-streamer` although it's on a private network I can see how one would like to exclude the schema from being sent to save on ingress/egress/networking cost, i.e.: like with a cloud provider.